src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -359,6 +359,8 @@ private static IncrementalStepRunReason AsStepState(EntryState inputState, Entry
                     (EntryState.Modified, EntryState.Cached) => IncrementalStepRunReason.Unchanged,
                     (EntryState.Cached, EntryState.Cached) => IncrementalStepRunReason.Cached,
                     (EntryState.Removed, EntryState.Removed) => IncrementalStepRunReason.Removed,
+                    (EntryState.Modified, EntryState.Removed) => IncrementalStepRunReason.Removed,
```

Comment:
Why not combine this with the one above using “or"?

Reply:
@AraHaan can you please give feedback on htat in this PR: https://github.com/dotnet/roslyn/pull/61308 

this is just a port of that work into this branch to unblock this work